crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -477,8 +477,9 @@ impl<'db> TypeInferenceBuilder<'db> {
         for class in class_definitions {
             // (1) Check that the class does not have a cyclic definition
             if class.is_cyclically_defined(self.db) {
-                self.diagnostics.add(
-                    class.node(self.db).into(),
+                self.diagnostics.add_if_definition_in_same_file(
```

Comment:
All diagnostics here should be guaranteed to be from the same file because `class.node` is from the current scope.

Reply:
Hm. You are probably right. I added it everywhere because the original error appeared with an example file without any class definitions at all:
```
from subprocess import Popen
```
Or do we consider the import of `Popen` to be a class definition in this file?

Reply:
> Or do we consider the import of `Popen` to be a class definition in this file?

The import of `Popen` constitutes a `Definition` and `Declaration` of the `Popen` symbol in this file, and the type of that `Declaration` is a `ClassLiteralType`!

Reply:
Ah, okay. Thanks! I reverted the check for the local classes.

Reply:
I'm still not sure we should be looking at any classes that originate from other modules in this iterator (https://github.com/astral-sh/ruff/pull/14337#discussion_r1842177196)

Reply:
No, I think we need that check here. I added an additional test to check for this. If we don't check this here, we will get an additional diagnostic at the line of the import (with a wrong `range`).

Reply:
I think the problem is the `self.types.declarations.values()`. We only want to iterate over classes that were declared in this file. Instead, it iterates over all declarations that have a class type, that includes imports and possibly others. 

We should adopt the filter above to filter out classes from other scopes AND only collect unique literals.

Reply:
Or a better way might be to filter based on the `declaration`s key to assert that it is a class definition

Reply:
Something like this

```
let class_definitions = self
            .types
            .declarations
            .iter()
            .filter_map(|(definition, ty)| {
                if matches!(definition.kind(self.db), DefinitionKind::Class(_)) {
                    ty.into_class_literal()
                } else {
                    None
                }
            })
            .map(|class_ty| class_ty.class);
```

Reply:
Sorry for keeping commenting. A related improvement would be that the iterator also returns the `node` so that we can replace the `class.node` calls (they're not expensive but getting them directly from the definition makes it more explicit that they are guaranteed to be from the same scope)

Reply:
> Something like this

Saw this too late, but came up with the exact same code. Except for a name difference (now changed to yours).





> A related improvement would be that the iterator also returns the `node`

Done.